sys/include/net/ieee802154.h

```diff
@@ -550,6 +550,16 @@ static inline uint8_t ieee802154_dbm_to_rssi(int16_t dbm)
     return val - IEEE802154_RADIO_RSSI_OFFSET;
 }
 
+/**
+ * @brief Get the symbol time for a given PHY channel.
+ *
+ * @param[in] channel_page channel page
+ * @param[in] channel_num channel number
+ *
+ * @return the symbol time in us
+ */
+uint32_t ieee802154_get_symbol_time(uint8_t channel_page, uint16_t channel_num);
```

Comment:
Is channel page really enough to determine the symbol time?
If I understand it correctly, all 802.15.4g-2012 modulations are mapped to page 9 & 10

![image](https://user-images.githubusercontent.com/1301112/214613943-c27301de-aa7b-4d20-a0b3-a0c524d0f9b4.png)

so channel page gives you no additional info. And for LoRa there isn't even a channel page AFAIK.

I think we won't get around a custom struct here.

```C

typedef struct {
    uint8_t rate_mode;
} ieee802154_oqpsk_params;

typedef struct {
    uint8_t chips;
    uint8_t rate_mode;
} ieee802154_mr_oqpsk_params;

…

uint32_t ieee802154_get_symbol_time(ieee802154_phy_mode_t phy, const void *params);
```

Reply:
>Is channel page really enough to determine the symbol time?

Not really but helps to identify which is the PHY used at the moment, for MR-FSK its a bit more complex since you need to have sort of a database of the standard defined PHY configurations and also the generic ones for page 10 that the standard allows to specify.

I think it's best to remove the `ieee802154_phy_mode_t` enum and use the channel page for that. However, I think there's still a need for a MR PHY modes enum since those share the page 9.

Reply:
Which reminds me we need a way to handle the special cases of pages 9 and 10. Not sure there's a need for page 10 but could be useful.

Reply:
But this is only internal, we don't have to use channel page if it complicates things.
Using our PHY type enum allows to also use this for non-IEEE 802.15.4 PHYs (and lets us use any mode combination, also the non-standard high-data rate modes implemented on some classic 802.15.4 radios).

Reply:
> And for LoRa there isn't even a channel page AFAIK.

There's no channel page for LoRa, but there are already use cases for IEEE 802.15.4 over LoRa. We could just pick a custom channel page that doesn't conflict with the standard.

> I think we won't get around a custom struct here.

Is there a way to abstract this information into a set of common input variables? Otherwise I'm afraid it will be really hard to write upper layer code :/.

> So we might need two functions

Sure, it makes sense. I would still keep the proposed `ieee802154_get_symbol_time` function though because slotted MAC layers use the symbol time as common multiple for timers.

Reply:
>  think it's best to remove the ieee802154_phy_mode_t enum and use the channel page for that. However, I think there's still a need for a MR PHY modes enum since those share the page 9.

We can do that.

Reply:
> Is there a way to abstract this information into a set of common input variables? Otherwise I'm afraid it will be really hard to write upper layer code :/.

AFAIS it should be possible to represent these channel settings with 5 variables:
1. Channel page
2. Channel number / frequency
3. Modulation scheme
4. Modulation mode
5. Mode options.

The `ieee802154_radio_config_phy` functions receives a structure as argument, which was meant to be used for PHYs that require more than a channel page and channel number. If we extend this struct, we should be able to configure any PHY state regardless of the modulation.
We can also reuse the same structure in the SubMAC descriptor, so we can just reference it when calling `ieee802154_radio_config_phy`.

Reply:
> Otherwise I'm afraid it will be really hard to write upper layer code :/

Hm you think so?
I didn't fully flesh this out yet, but wouldn't it just be a

```C
typedef struct {
    ieee802154_phy_mode_t pyh;
    union {
#ifdef MODULE_NETDEV_IEEE802154_OQPSK
        struct {
            bool subghz;
            bool high_rate;
        } oqpsk_params;
#endif
#ifdef MODULE_NETDEV_IEEE802154_MR_OQPSK
        struct {
            uint8_t chips;
            uint8_t rate_mode;
        } mr_oqpsk_params;
#endif
        …

    } params;
} ieee802154_submac_phy_params_t;
```

and a 

```patch
--- a/sys/include/net/ieee802154/submac.h
+++ b/sys/include/net/ieee802154/submac.h
@@ -206,6 +206,7 @@ struct ieee802154_submac {
     ieee802154_fsm_state_t fsm_state;    /**< State of the SubMAC */
     ieee802154_phy_mode_t phy_mode;     /**< IEEE 802.15.4 PHY mode */
     const iolist_t *psdu;               /**< stores the current PSDU */
+    ieee802154_submac_phy_params_t phy;
 };
 
 /**
```

Maybe I should try a PR to see where this goes :sweat_smile:

Reply:
Ah, you are right. Probably is not that hard to go that way, since we will eventually update the `config_phy` function of the Radio HAL anyway.

We could then kill 2 birds with one stone with this:
```diff
diff --git a/sys/include/net/ieee802154/radio.h b/sys/include/net/ieee802154/radio.h
index c8f71afa90..aa3a6e5d6e 100644
--- a/sys/include/net/ieee802154/radio.h
+++ b/sys/include/net/ieee802154/radio.h
@@ -453,11 +453,38 @@ typedef enum {
  * @brief Holder of the PHY configuration
  */
 typedef struct {
-    ieee802154_phy_mode_t phy_mode; /**< IEEE802.15.4 PHY mode */
+    typedef union {
+#ifdef MODULE_IEEE802154_OQPSK
+        struct {
+            bool subghz;
+            bool high_rate;
+        } oqpsk_params;
+#endif
+#ifdef MODULE_IEEE802154_MR_OQPSK
+        struct {
+            uint8_t chips;
+            uint8_t rate_mode;
+        } mr_oqpsk_params;
+#endif
+#ifdef MODULE_IEEE802154_MR_ODFM
+        struct {
+            uint8_t option;
+            uint8_t msc;
+        } mr_odfm_params;
+#endif
+#ifdef MODULE_IEEE802154_MR_FSK
+        struct {
+            uint8_t rate;
+            uint8_t mod_idx
+            uint8_t mod_order;
+            uint8_t preamble_len;
+            bool fec;
+        } mr_fsk_params;
+#endif
     uint16_t channel;               /**< IEEE802.15.4 channel number */
     uint8_t page;                   /**< IEEE802.15.4 channel page */
     int8_t pow;                     /**< TX power in dBm */
-} ieee802154_phy_conf_t;
+} ieee802154_phy_pib_t;
 
 /**
  * @brief IEEE 802.15.4 radio operations
@@ -686,7 +713,7 @@ struct ieee802154_radio_ops {
      * @return -EINVAL  if the configuration is not valid for the device.
      * @return <0       error, return value is negative errno indicating the cause.
      */
-    int (*config_phy)(ieee802154_dev_t *dev, const ieee802154_phy_conf_t *conf);
+    int (*config_phy)(ieee802154_dev_t *dev, const ieee802154_phy_pib_t *conf);
 
     /**
      * @brief Set number of frame retransmissions
@@ -915,7 +942,7 @@ static inline int ieee802154_radio_set_cca_mode(ieee802154_dev_t *dev,
  * @return result of @ref ieee802154_radio_ops::config_phy
  */
 static inline int ieee802154_radio_config_phy(ieee802154_dev_t *dev,
-                                              const ieee802154_phy_conf_t *conf)
+                                              const ieee802154_phy_pib_t *conf)
 {
     return dev->driver->config_phy(dev, conf);
 }
diff --git a/sys/include/net/ieee802154/submac.h b/sys/include/net/ieee802154/submac.h
index 3890ab7579..42f0d934f0 100644
--- a/sys/include/net/ieee802154/submac.h
+++ b/sys/include/net/ieee802154/submac.h
@@ -196,15 +196,13 @@ struct ieee802154_submac {
     ieee802154_csma_be_t be;            /**< CSMA-CA backoff exponent params */
     bool wait_for_ack;                  /**< SubMAC is waiting for an ACK frame */
     uint16_t panid;                     /**< IEEE 802.15.4 PAN ID */
-    uint16_t channel_num;               /**< IEEE 802.15.4 channel number */
-    uint8_t channel_page;               /**< IEEE 802.15.4 channel page */
     uint8_t retrans;                    /**< current number of retransmissions */
     uint8_t csma_retries_nb;            /**< current number of CSMA-CA retries */
     uint8_t backoff_mask;               /**< internal value used for random backoff calculation */
     uint8_t csma_retries;               /**< maximum number of CSMA-CA retries */
     int8_t tx_pow;                      /**< Transmission power (in dBm) */
-    ieee802154_fsm_state_t fsm_state;    /**< State of the SubMAC */
-    ieee802154_phy_mode_t phy_mode;     /**< IEEE 802.15.4 PHY mode */
+    ieee802154_fsm_state_t fsm_state;   /**< State of the SubMAC */
+    ieee802154_phy_pib_t pib;           /**< IEEE 802.15.4 Physical Information Base */
     const iolist_t *psdu;               /**< stores the current PSDU */
 };
 

```

We would need to adapt `ieee802154_set_phy_conf` though to make it more generic.

Reply:
please correct me if I'm wrong: do all IEEE 802.15.4 channel pages use a channel number to identify the channel?

Reply:
> do all IEEE 802.15.4 channel pages use a channel number to identify the channel?

Yes. (But I still don't understand what the channel pages are actually used for)

Reply:
I implemented the suggested approach in #19668

Reply:
> > do all IEEE 802.15.4 channel pages use a channel number to identify the channel?
> 
> Yes. (But I still don't understand what the channel pages are actually used for)

isn't the page used to define how to fit channels in some regulatory frequency's  and the channel then identifys wihsch of them is used 

eg 350MHz to 358MHz fits 7 1MHZ channels but you might want to evenly spread 5  to have some wiggle room left an right of each channel

another modulation might be slimmer and you can fit 6 channels